lib/ui/text/paragraph_builder.cc

```diff
@@ -417,19 +417,34 @@ void ParagraphBuilder::pop() {
   }
 }
 
-void ParagraphBuilder::addText(const std::string& text) {
+Dart_Handle ParagraphBuilder::addText(const std::u16string& text) {
+  if (text.empty())
+    return Dart_Null();
+
+  // Use ICU to validate the UTF-16 input.  Calling u_strToUTF8 with a null
+  // output buffer will return U_BUFFER_OVERFLOW_ERROR if the input is well
+  // formed.
+  const UChar* text_ptr = reinterpret_cast<const UChar*>(text.data());
+  UErrorCode error_code;
+  u_strToUTF8(nullptr, 0, nullptr, text_ptr, text.size(), &error_code);
+  if (error_code != U_BUFFER_OVERFLOW_ERROR)
```

Comment:
Check for `U_ZERO_ERROR` here?

Reply:
Any code other than U_BUFFER_OVERFLOW_ERROR indicates that the input was not valid UTF-16

Reply:
Thanks for clarifying.